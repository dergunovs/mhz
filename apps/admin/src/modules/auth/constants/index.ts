export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
