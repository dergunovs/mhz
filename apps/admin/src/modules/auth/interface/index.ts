export interface ILoginFormData {
  email: string;
  password: string;
}
