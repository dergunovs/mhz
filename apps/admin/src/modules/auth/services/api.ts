import { api, setAuth } from 'mhz-helpers';

import {
  API_AUTH_CHECK,
  API_AUTH_LOGIN,
  API_AUTH_SETUP,
  ILoginData,
  ISignUpData,
  IBaseReply,
  IUserToken,
} from 'mhz-contracts';

export async function checkAuthApi() {
  const { data } = await api.get<IBaseReply>(API_AUTH_CHECK);

  if (data.message) setAuth(true);

  return data;
}
export async function setupApi(formData: ISignUpData) {
  const { data } = await api.post<IBaseReply>(API_AUTH_SETUP, formData);

  return data;
}

export async function loginApi(formData: ILoginData) {
  const { data } = await api.post<IUserToken>(API_AUTH_LOGIN, formData);

  return data;
}
