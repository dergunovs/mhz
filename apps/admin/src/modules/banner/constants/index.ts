export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
