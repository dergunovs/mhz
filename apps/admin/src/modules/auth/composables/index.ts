import { readonly, ref } from 'vue';
import { useRouter } from 'vue-router';

import { setAuthHeader, deleteAuthHeader } from '@/common/services/api';
import { URL_LOGIN } from '@/auth/constants';
import { URL_MAIN } from '@/common/constants';

const isAuthLocal = ref(false);

export const isAuth = readonly(isAuthLocal);

export function setAuth(auth: boolean) {
  isAuthLocal.value = auth;
}

export function logout() {
  deleteCookieToken();
  deleteAuthHeader();
  window.location.href = `${URL_LOGIN}?logout=1`;
}

export function setCookieToken(token: string) {
  document.cookie = `mhz_token=${token};Secure;samesite=strict;`;
}

export function getCookieToken(): string | undefined {
  const { mhz_token } = Object.fromEntries(document.cookie.split('; ').map((v) => v.split('=')));

  return mhz_token;
}

export function deleteCookieToken() {
  document.cookie = `mhz_token=;expires=${new Date(0).toUTCString()}`;
}

export function useAuth() {
  const router = useRouter();

  function login(token: string) {
    setCookieToken(token);
    setAuthHeader(token);
    setAuth(true);

    router.push(URL_MAIN);
  }

  return {
    login,
  };
}
