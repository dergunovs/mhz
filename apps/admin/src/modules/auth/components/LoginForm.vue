<template>
  <Form @submit="submit" layout="vertical">
    <FormItem label="E-mail" v-bind="validateInfos.email" required>
      <Input v-model:value="formData.email" />
    </FormItem>

    <FormItem label="Пароль" v-bind="validateInfos.password" required>
      <InputPassword v-model:value="formData.password" />
    </FormItem>

    <Button htmlType="submit" type="primary">Войти</Button>
  </Form>
</template>

<script setup lang="ts">
import { ref } from 'vue';

import { Button, Form, FormItem, Input, InputPassword } from 'ant-design-vue';
import type { Rule } from 'ant-design-vue/es/form';

import { postLogin } from '@/auth/services';
import { useAuth } from '@/auth/composables';

const { login } = useAuth();

const formData = ref({
  email: '',
  password: '',
});

const rules = ref<{ [key: string]: Rule[] }>({
  email: [{ required: true, type: 'email' }],
  password: [{ required: true }],
});

const useForm = Form.useForm;

const { validate, validateInfos } = useForm(formData, rules);

const { mutate } = postLogin(formData, {
  onSuccess: (token?: string) => {
    if (token) login(token);
  },
});

async function submit() {
  try {
    await validate();
    mutate();
  } catch (error) {
    alert(`Исправьте ошибки валидации`);
  }
}
</script>
