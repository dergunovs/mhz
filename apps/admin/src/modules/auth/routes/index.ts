import { URL_LOGIN } from '@/auth/constants';

export const authRoutes = [{ path: URL_LOGIN, name: 'Login', component: () => import('@/auth/pages/LoginPage.vue') }];
