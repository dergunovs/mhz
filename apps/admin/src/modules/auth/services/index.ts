import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
