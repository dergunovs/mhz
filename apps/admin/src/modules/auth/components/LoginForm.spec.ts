import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { ILoginData, IUserToken } from 'mhz-contracts';
import { toast } from 'mhz-ui';
import * as helpers from 'mhz-helpers';

import LoginForm from './LoginForm.vue';

import { TOKEN_NAME } from '@/auth/constants';
import { URL_MAIN } from '@/common/constants';

import { wrapperFactory, mockMutationReply } from '@/common/test';
import * as authServices from '@/auth/services';

const EMAIL = 'a@b.ru';
const PASSWORD = 'qwerty';
const ROLE = 'manager';
const TOKEN = 'b87dfnyte97bhrevber9vu9e';
const ID = '97fdb9eubhe';

let onSuccessLogin: (data: IUserToken) => void;
const spyMutateLogin = vi.fn();

vi.spyOn(authServices, 'login').mockImplementation((options: { onSuccess?: (data: IUserToken) => void }) => {
  if (options.onSuccess) onSuccessLogin = options.onSuccess;

  return mockMutationReply<IUserToken, ILoginData>(spyMutateLogin);
});

const spyAuth = vi.fn();

vi.spyOn(helpers, 'useAuth').mockReturnValue({ auth: spyAuth, redirectIfAuth: () => undefined });

const spySetAuthHeaders = vi.spyOn(helpers, 'setAuthHeader');

const spyToastSuccess = vi.spyOn(toast, 'success');

let wrapper: VueWrapper;

const loginFormEmail = '[data-test="login-form-email"]';
const loginFormPassword = '[data-test="login-form-password"]';
const loginFormButton = '[data-test="login-form-button"]';

beforeEach(() => {
  wrapper = wrapperFactory(LoginForm, {});
});

enableAutoUnmount(afterEach);

describe('LoginForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(LoginForm)).toBeTruthy();
  });

  it('handles login by form submit', async () => {
    await wrapper.findComponent(loginFormEmail).setValue(EMAIL);
    await wrapper.findComponent(loginFormPassword).setValue(PASSWORD);

    await wrapper.find(loginFormButton).trigger('click');

    expect(spyMutateLogin).toBeCalledTimes(1);
    expect(spyMutateLogin).toBeCalledWith({ email: EMAIL, password: PASSWORD, role: ROLE });

    onSuccessLogin({ _id: ID, email: EMAIL, role: ROLE, token: TOKEN });

    expect(spyAuth).toBeCalledTimes(1);
    expect(spyAuth).toBeCalledWith(TOKEN, URL_MAIN, spySetAuthHeaders, TOKEN_NAME);

    expect(spyToastSuccess).toBeCalledTimes(1);
  });
});
