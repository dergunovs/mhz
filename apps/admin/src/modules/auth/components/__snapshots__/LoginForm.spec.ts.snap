// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > matches snapshot 1`] = `
"<div>
  <anonymous-stub></anonymous-stub>
  <h2>Login</h2>
  <form class="_form_b133dc" data-test="login-form">
    <fieldset label="E-mail" isrequired=""><input type="text" modelvalue="" isfocus="" data-test="login-form-email"></fieldset>
    <fieldset label="Password" isrequired=""><input type="password" modelvalue="" data-test="login-form-password"></fieldset><button type="submit">Submit</button>
  </form>
</div>"
`;
